polymorphic.h

```diff
@@ -132,99 +161,108 @@ class polymorphic {
+      u_traits::deallocate(u_alloc, mem, 1);
       throw;
     }
   }
 
   polymorphic(const polymorphic& other) : alloc_(other.alloc_) {
-    assert(other.cb_ != nullptr);  // LCOV_EXCL_LINE
-    cb_ = other.cb_->clone(alloc_);
+    assert(other.p_ != nullptr);  // LCOV_EXCL_LINE
+    cb_ = other.cb_;
+    p_ = cb_->clone(alloc_, *other.p_);
   }
 
   polymorphic(std::allocator_arg_t, const A& alloc, const polymorphic& other) : alloc_(alloc) {
-    assert(other.cb_ != nullptr);  // LCOV_EXCL_LINE
-    cb_ = other.cb_->clone(alloc_);
+    assert(other.p_ != nullptr);  // LCOV_EXCL_LINE
+    cb_ = other.cb_;
+    p_ = cb_->clone(alloc_, *other.p_);
   }
 
   polymorphic(polymorphic&& other) noexcept : alloc_(std::move(other.alloc_)) {
-    assert(other.cb_ != nullptr);  // LCOV_EXCL_LINE
-    cb_ = std::exchange(other.cb_, nullptr);
+    assert(other.p_ != nullptr);  // LCOV_EXCL_LINE
+    cb_ = other.cb_;
+    p_ = std::exchange(other.p_, nullptr);
   }
 
   polymorphic(std::allocator_arg_t, const A& alloc, polymorphic&& other) noexcept : alloc_(alloc){
-    assert(other.cb_ != nullptr);  // LCOV_EXCL_LINE
-    cb_ = std::exchange(other.cb_, nullptr);
+    assert(other.p_ != nullptr);  // LCOV_EXCL_LINE
+    cb_ = other.cb_;
+    p_ = std::exchange(other.p_, nullptr);
   }
 
   ~polymorphic() { reset(); }
 
   polymorphic& operator=(const polymorphic& other) {
-    assert(other.cb_ != nullptr);  // LCOV_EXCL_LINE
+    assert(other.p_ != nullptr);  // LCOV_EXCL_LINE
     polymorphic tmp(other);
     swap(tmp);
     return *this;
   }
 
   polymorphic& operator=(polymorphic&& other) noexcept {
-    assert(other.cb_ != nullptr);  // LCOV_EXCL_LINE
+    assert(other.p_ != nullptr);  // LCOV_EXCL_LINE
     reset();
     alloc_ = std::move(other.alloc_);
-    cb_ = std::exchange(other.cb_, nullptr);
+    cb_ = other.cb_;
+    p_ = std::exchange(other.p_, nullptr);
     return *this;
   }
 
   constexpr T* operator->() noexcept {
-    assert(cb_ != nullptr);  // LCOV_EXCL_LINE
-    return cb_->p_;
+    assert(p_ != nullptr);  // LCOV_EXCL_LINE
+    return p_;
   }
 
   constexpr const T* operator->() const noexcept {
-    assert(cb_ != nullptr);  // LCOV_EXCL_LINE
-    return cb_->p_;
+    assert(p_ != nullptr);  // LCOV_EXCL_LINE
+    return p_;
   }
 
   constexpr T& operator*() noexcept {
-    assert(cb_ != nullptr);  // LCOV_EXCL_LINE
-    return *cb_->p_;
+    assert(p_ != nullptr);  // LCOV_EXCL_LINE
+    return *p_;
   }
 
   constexpr const T& operator*() const noexcept {
-    assert(cb_ != nullptr);  // LCOV_EXCL_LINE
-    return *cb_->p_;
+    assert(p_ != nullptr);  // LCOV_EXCL_LINE
+    return *p_;
   }
 
   constexpr bool valueless_after_move() const noexcept {
-    return cb_ == nullptr;
+    return p_ == nullptr;
   }
 
   constexpr void swap(polymorphic& other) noexcept {
-    assert(other.cb_ != nullptr);  // LCOV_EXCL_LINE
+    // NOTE: I don't think these asserts should be here. Even if one is valueless it should be swappable
```

Comment:
I think that these asserts are correct. We require that move-from is the only way to get into a valueless state after which any access to the object other than to assign to it or destroy it is invalid.

Reply:
I would argue that swap is just a double assign and should be allowed. The litmus test would be whether a variant in valueless_by_exception state is swappable.